Github Webhooks: Add parameter to hide body on updated issues/PRs

Pritesh-30 opened the pull request:
Fixes part of #32323.

This PR adds support for a new `compact_edit` URL option in the GitHub webhook
integration to hide body content for **edited issues and pull requests**.

Added the `compact_edit` URL option to `integrations.py` and updated the gitHubs integration documentation

**How changes were tested:**
- Ran `./tools/test-backend zerver/webhooks/github` all test suites pass

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
When no compact_edit or compact_edit=false
<img width="1094" height="162" alt="image" src="https://github.com/user-attachments/assets/abcbb785-9eba-45c7-a615-a9c7334409be" />
when compact_edit=true
<img width="1121" height="74" alt="image" src="https://github.com/user-attachments/assets/fb725fb2-d471-4e4b-a2a7-3a7e40e3652e" />
When PR is edited
<img width="1124" height="138" alt="image" src="https://github.com/user-attachments/assets/4e6a4475-fe4e-4965-a64a-df7470302e3c" />


<img width="761" height="827" alt="image" src="https://github.com/user-attachments/assets/882e2f3a-22ef-44a0-9a3a-9bc09c2d2bc9" />
<img width="752" height="824" alt="image" src="https://github.com/user-attachments/assets/2960c261-be11-426a-ae5b-6b6df80915c9" />


<img width="1499" height="789" alt="image" src="https://github.com/user-attachments/assets/5e7ace6d-f852-4768-a3c9-c85d6a3a9872" />


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Pritesh-30 commented:
> "Hi @Pritesh-30, great work on setting up the fixtures!
> 
> I’ve been working on the second half of this issue locally—specifically splitting the opened/edited logic and implementing the compact_edit=true parameter. I have it working with 130/130 tests passing, including a new test case for the compact mode.
> 
> Since you've already started this PR, would you like me to share my logic here so you can include it? Or if you're busy, I can open a new PR that builds on top of your fixtures. Let me know what works best for you!"

@Lakshyalamba Thanks for the update!
I haven’t started working on the second part yet since I’m currently busy with another issue. If you already have the logic implemented and all test suites are passing, it would be great if you could take it forward.
Yes, please feel free to build on top of the fixtures I added and open a PR. When doing so, make sure to reference the commit where the fixtures were introduced, following the guidelines here: https://zulip.readthedocs.io/en/latest/contributing/continuing-unfinished-work.html

Let me know if you have any questions I can help you

Pritesh-30 committed:
webhooks/github: Add compact_edit for edited issues and PRs.

Adds a new URL parameter for the GitHub integration to suppress body content
in edited issue and pull request events.

Fixes part of #32323.

Niloth-p commented on zerver/webhooks/github/view.py:
```diff
@@ -165,28 +167,56 @@ def get_member_body(helper: Helper) -> str:
+        include_body=not helper.compact_edit,
+    )
+
+
```
```suggestion
    show_body_map = {
        "edited": not helper.compact_edit,
        "assigned": False,
        "unassigned": False,
    }
    message = ... if show_body_map.get(action, True) else None
```

Why break them into 3 functions just to set the `include_body` value?

Niloth-p reviewed (commented):
Thanks for the PR!

1. The first commit is also included in #37183. Since both the PRs are dependent on that commit, we could split that out as a separate PR with both of you as authors. You should discuss to figure out who wants to open that PR. Make sure to add separate fixtures for title edits and body edits like in that PR. Relevant comment: https://github.com/zulip/zulip/pull/37183#discussion_r2699485666
2. Why is this PR claiming to fix the issue? This PR does not add support for hiding the body on updated PRs, discussions, discussion comments, etc. It is fine to limit this PR just to the issue updates, but make sure to edit the PR description (and title too), to not auto-close the issue.
3. `compact_edit` needs to be added as a URL option in `integrations.py`. See the other URL options of GitHub for examples. Make sure to update the doc, and add a screenshot of the modal to the PR description. It would be nice to include PR events in the same PR so that we can use a generic description for that URL option in the modal.

Niloth-p commented on zerver/webhooks/github/doc.md:
```diff
@@ -35,4 +35,10 @@ Get GitHub notifications in Zulip!
 
 {!webhooks-url-specification.md!}
 
+### Configuration options
```
I had previously mentioned "See the other URL options of GitHub for examples." in my past review. So, why is this adding a new section to the doc when GitHub already has configuration options?

Niloth-p reviewed (commented):
Thanks for the updates!

I believe we should make `compact_edit` true by default. #37255 adds support for the corresponding checkbox. Until that gets merged, you could work on adding support for the PRs.

Niloth-p commented:
I'm waiting for #37255 to be done.

Niloth-p commented:
Now that #38854 has been merged, this PR is unblocked. Use the new `enabled_checkbox`.

Pritesh-30 commented:
Updates
- Used  new `enabled_checkbox` to make `compact_edit` true  by default
- Updated the PR description with new screenshots

Niloth-p commented on zerver/webhooks/github/tests.py:
````diff
@@ -163,6 +163,14 @@ def test_issues_edited_title(self) -> None:
         expected_message = "Pritesh-30 edited [issue #6](https://github.com/Pritesh-30/test-repo/issues/6):\n\n``` quote\nThe body of the issue is edited.\n```"
         self.check_webhook("issues__edited_title", expected_topic_name, expected_message)
 
+    def test_issue_edited_compact(self) -> None:
````
```suggestion
    def test_issues_edited_body_compact(self) -> None:
```

Niloth-p commented on zerver/webhooks/github/view.py:
```diff
@@ -176,16 +178,23 @@ def get_issue_body(helper: Helper) -> str:
+
+    message = (
+        issue["body"].tame(check_none_or(check_string)) if show_body_map.get(action, True) else None
+    )
```
How about we drop this map, and instead just use boolean checks.
```suggestion
	is_compact_edit = action == "edited" and helper.compact_edit
	is_assignment_action = action in ("assigned", "unassigned")
	message = None if (is_compact_edit or is_assignment_action) else issue["body"].tame(check_none_or(check_string))
```

Pritesh-30 replied:
Sure I will drop this map and use boolean checks

Niloth-p commented on zerver/webhooks/github/view.py:
```diff
@@ -1150,6 +1159,7 @@ def api_github_webhook(
     user_specified_topic: OptionalUserSpecifiedTopicStr = None,
     ignore_private_repositories: Json[bool] = False,
     include_repository_name: Json[bool] = False,
+    compact_edit: Json[bool] = False,
```
```suggestion
    compact_edit: Json[bool] = True,
```

Niloth-p commented on zerver/webhooks/github/tests.py:
```diff
@@ -453,6 +461,13 @@ def test_pull_request_edited_with_body_change(self) -> None:
         self.check_webhook("pull_request__edited_with_body_change", TOPIC_PR, expected_message)
 
+    def test_pull_request_edited_compact(self) -> None:
+        self.url = self.build_webhook_url(compact_edit="true")
```
Compact edit should be true by default. The tests need to be inverted.

Niloth-p reviewed (commented):
Wow, that was super quick. Thanks for the update, but the default value of `compact_edit` should be True.

Pritesh-30 commented:
> Wow, that was super quick. Thanks for the update, but the default value of `compact_edit` should be True.

@Niloth-p  Thanks for the review! I was following the updates on your PR. So I had started working on the changes.
Updates made
- Changed test name `test_issue_edited_compact` --> `test_issues_edited_body_compact`
- Changed test name `test_pull_request_edited_compact` --> `test_pull_request_edited__with_body_change_compact`
- Dropped `show_body_map` and now use boolean checks
- Kept default value of `compact_edit` as `True`
- Updated tests `test_issues_edited_body`,`test_issues_edited_title` and `test_pull_request_edited_with_body_change`  to use `compact_edit=false`

Pritesh-30 commented:
@Niloth-p Updated the docs

laurynmm commented:
@apoorvapendse would you be up for picking up maintainer review here?

Not sure if the "changes requested" label should be removed. @Pritesh-30 can you note where that was added and if the feedback was resolved?

Pritesh-30 commented:
It was added when this PR still contained two commits, one of which was the common fixtures commit that was also needed for Dhruv's PR. As part of the review, that commit was moved into a separate PR #37551, which has since been merged. The remaining feedback was also addressed by adding `compact_edit` as a URL option in integrations.py and making it enabled `True` by default. So I think it is safe to remove `changes requested` label

apoorvapendse commented on zerver/webhooks/github/tests.py:
````diff
@@ -450,9 +459,16 @@ def test_pull_request_edited_msg(self) -> None:
         expected_message = "cozyrohan edited [PR #1](https://github.com/cozyrohan/public-repo/pull/1):\n\n``` quote\nPR EDITED\n```"
         self.check_webhook("pull_request__edited_with_body_change", TOPIC_PR, expected_message)
 
+    def test_pull_request_edited__with_body_change_compact(self) -> None:
````
Let's get rid of the double underscore.

apoorvapendse commented on zerver/lib/integrations.py:
```diff
@@ -678,6 +678,11 @@ def is_enabled_in_catalog(self) -> bool:
             ),
+            WebhookUrlOption(
+                name="compact_edit",
+                label="Exclude issue and pull request descriptions on edit events",
```
Should we be calling this edit notifications, to be consistent with the other options and doc.md?

Pritesh-30 replied:
Yes I will update it to call `edit notifications`.

apoorvapendse commented on zerver/webhooks/github/tests.py:
```diff
@@ -153,16 +153,25 @@ def test_fork_msg(self) -> None:
         self.check_webhook("fork", TOPIC_REPO, expected_message)
 
     def test_issues_edited_body(self) -> None:
+        self.url = self.build_webhook_url(compact_edit="false")
```
Was the decision to switch to a compact version by default discussed somewhere, shouldn't the non-compact version be the default, since that is what we had before?

Wouldn't [this](https://github.com/zulip/zulip/issues/32323#issuecomment-2486736044) version be nicer compared to switching everything completely for everyone?

> Would excluding notices related to PRs with certain authors, like dependabot, be a better way to achieve your goals?

Pritesh-30 replied:
Regarding filtering by author/event-type like Dependabot, that was discussed in the original issue. The reporter mentioned that those notifications are still useful, and anyone not interested can simply mute the PR's topic. That's why I implemented it as a configurable URL option instead.

Pritesh-30 replied:
And about keeping the compact version default `True` there wasn't any concrete discussion. However, @Niloth-p had explicitly mentioned during an earlier review that `compact_edit` should be kept `True` by default.

apoorvapendse replied:
OK, thanks for the context.

apoorvapendse reviewed (commented):
Thanks, left some comments.

apoorvapendse reviewed (approved):
LGTM, assigning @Niloth-p for integration review.

laurynmm commented:
I'll do the integration review here, though @Niloth-p if you do have time to look at these changes, feel free to post any review comments.

Niloth-p commented:
Thank you for the review and the good catches, @apoorvapendse.

@laurynmm Thank you. My previous suggestion has been addressed, so assuming no unaccounted-for changes, I don't expect to have new comments.

laurynmm commented:
@Pritesh-30 - Thanks for working on this issue! I went ahead and pulled down these changes to make a few revisions that I'll outline below.

Also, started a CZO discussion for a few more questions/thoughts I have before merging these changes: [#integrations > GitHub integration - compact edited notifications](https://chat.zulip.org/#narrow/channel/127-integrations/topic/GitHub.20integration.20-.20compact.20edited.20notifications/with/2511375)

EDIT: CI failure is an unrelated puppeteer test flake.

---

My updates:
- Changed the URL field to `compact_edit_format`, because `compact_edit` seemed that it could be a bit confusing, like it was a type of edit vs the format of the notification message.
- Change the strings in the URL modal and doc to be "Use a compact message format for edited issue and pull request descriptions", because I felt that explanation went better with the URL field name.
- Revised the commit message for my changes.

---

Updated screenshots:

<details>
<summary>Integration doc</summary>

<img width="1762" height="1004" alt="Screenshot From 2026-08-18 18-49-43" src="https://github.com/user-attachments/assets/2e0bb575-05f9-46ac-a054-835ce55c079c" />
</details>
<details>
<summary>URL modal in web app</summary>

<img width="1224" height="1458" alt="Screenshot From 2026-08-18 18-48-48" src="https://github.com/user-attachments/assets/be464424-5ed6-45f8-b9e2-b476082cd992" />
</details>
